### ISSUE TITLE
Make Exceptionless.Json private or embed shared code as DLL resources in binaries

### ISSUE THREAD

**Issue by leidegre** [exceptionless/Exceptionless.Net#165]
I get why you'd want to lock down what versions of what is being used, not pollute solutions with unnecessary dependencies and I admire the effort to do so. But I have a vanity request/suggestion.

It might be useful to you it might not be but I thought I'd put it up for discussion, I can help with this as well if you think it's a reasonable idea.

I'm annoyed a little bit by the `Exceptionless.Json` namespace as it pops up every now and then in our solutions. I might have misunderstood but there's no reason for any non-exceptionless code to use that namespace, yes?

I have a couple of suggestions.

- I could write a tool that would change the default accessibility for all types in a specific namespace to internal. InternalsVisisbleTo can be used to then share these types within the exceptionless ecosystem.

- Build shared exceptionless code as a separate DLL and embed this DLL as a resource, then load this resource from the assembly manifest when needed. It can be done with just a few lines of code and doesn't require fancy post-build steps. _As a bonus, you can also opt to compress DLLs for a smaller footprint._

With what I've laid out so far, is this something you might consider changing?

**Comment by niemyjski** [comment]
We can't make it private because then you can't use attributes to do exclusions for serialization (We could probably introduce an interface of our own). Perhaps we can make some more of it internal. 

How are you adding exceptionless as a dependency? We only add it to the entry points of our applications so it's not referenced by any core libraries.

@ejsmith thoughts?

**Comment by ejsmith** [comment]
I’ve run into the same issues where I accidentally use the wrong namespace for Json stuff. I think we could probably make most of the classes internal and then maybe even change the name of some of the attributes for doing things like JsonIgnore. It would be a breaking change, but it may be worth it.

**Comment by leidegre** [comment]
It would be a welcome change, it's a minor inconvenience, as you say it should only show up in entry points but every now and then it crops up as an accidental mistake. Additionally, a tooling like ReSharper will pick up the type in these DLLs and offer it as a suggestion from time to time.

**Comment by ejsmith** [comment]
@leidegre would you be able to make this change and submit a PR?

**Comment by ejsmith** [comment]
I think we could actually internalize the whole thing and then just tell people to either just use `DataMember` or `DataMemberIgnore` attributes to control simple name changes and ignoring members or to use their own serializer.

**Comment by leidegre** [comment]
@ejsmith if that is the direction you wanna go with, I'm all for it. If you need me to write a code mod and just change everything `Exceptionless.Json` from public to internal, I can certainly do that.

**Comment by ejsmith** [comment]
@leidegre yeah I think that is what we should do. There is already a script that updates the source here: 

https://github.com/exceptionless/Exceptionless.Net/blob/master/src/Exceptionless/update-json.ps1

Would definitely appreciate any help!

**Comment by leidegre** [comment]
OK, good, I'll build upon that.

**Comment by niemyjski** [comment]
@leidegre thanks! I just ran into this on Friday and this would be a very welcomed change and greatly appreciated! If you need any assistance please let us know.

**Comment by niemyjski** [comment]
@leidegre did you have any luck on this? Let us know if there is anything we can do to help

**Comment by leidegre** [comment]
Actually yes but I've only had a couple of minutes to work on this.

I added this line to the PowerShell script

    $c = $c -replace 'public( (?:static|abstract))? (class|struct|interface|enum)','internal$1 $2'

Which I think gets us 90% there, haven't verified everything though, have to compile things and possibly add some internals visible to attributes. But that's as I've made it given the little time I've had to spare.

**Comment by niemyjski** [comment]
No worries, just wanted to see what is going into the next release :)

Thanks
-Blake Niemyjski

**Comment by yang-xiaodong** [comment]
I have submit a PR to fix this problem. #199

**Comment by niemyjski** [comment]
Thanks for the pr, I noticed the powershell script wasn't updated so it's going to be really hard for us to update as well as extra attributes were removed (any reason for this, we try to stay as close to the json.net release as possible with any modifications coming via the powershell script so it's easy to maintain.
